machine_learning/loss_functions/categorical_cross_entropy.py

```diff
@@ -0,0 +1,85 @@
+"""
+Categorical Cross-Entropy Loss
+
+This function calculates the Categorical Cross-Entropy Loss between true class
+labels and predicted class probabilities.
+
+Formula:
+Categorical Cross-Entropy Loss = -Σ(y_true * log(y_pred))
+
+Resources:
+- [Wikipedia - Cross entropy](https://en.wikipedia.org/wiki/Cross_entropy)
+"""
+
+import numpy as np
+
+
+def categorical_cross_entropy(
+    y_true: np.ndarray, y_pred: np.ndarray, epsilon: float = 1e-15
+) -> float:
+    """
+    Calculate Categorical Cross-Entropy Loss between true class labels and
+    predicted class probabilities.
+
+    Parameters:
+    - y_true: True class labels (one-hot encoded) as a NumPy array.
+    - y_pred: Predicted class probabilities as a NumPy array.
+    - epsilon: Small constant to avoid numerical instability.
+
+    Returns:
+    - ce_loss: Categorical Cross-Entropy Loss as a floating-point number.
+
+    Example:
+    >>> true_labels = np.array([[1, 0, 0], [0, 1, 0], [0, 0, 1]])
+    >>> pred_probs = np.array([[0.9, 0.1, 0.0], [0.2, 0.7, 0.1], [0.0, 0.1, 0.9]])
+    >>> categorical_cross_entropy(true_labels, pred_probs)
+    0.567395975254385
+
+    >>> y_true = np.array([[1, 0], [0, 1]])
+    >>> y_pred = np.array([[0.9, 0.1, 0.0], [0.2, 0.7, 0.1]])
+    >>> categorical_cross_entropy(y_true, y_pred)
+    Traceback (most recent call last):
+        ...
+    ValueError: Input arrays must have the same shape.
+
+    >>> y_true = np.array([[2, 0, 1], [1, 0, 0]])
+    >>> y_pred = np.array([[0.9, 0.1, 0.0], [0.2, 0.7, 0.1]])
+    >>> categorical_cross_entropy(y_true, y_pred)
+    Traceback (most recent call last):
+        ...
+    ValueError: y_true must be one-hot encoded.
+
+    >>> y_true = np.array([[1, 0, 1], [1, 0, 0]])
+    >>> y_pred = np.array([[0.9, 0.1, 0.0], [0.2, 0.7, 0.1]])
+    >>> categorical_cross_entropy(y_true, y_pred)
+    Traceback (most recent call last):
+        ...
+    ValueError: y_true must be one-hot encoded.
+
+    >>> y_true = np.array([[1, 0, 0], [0, 1, 0]])
+    >>> y_pred = np.array([[0.9, 0.1, 0.1], [0.2, 0.7, 0.1]])
+    >>> categorical_cross_entropy(y_true, y_pred)
+    Traceback (most recent call last):
+        ...
+    ValueError: Predicted probabilities must sum to approximately 1.
+    """
+    if y_true.shape != y_pred.shape:
+        raise ValueError("Input arrays must have the same shape.")
+
+    if not np.array_equal(y_true[y_true != 0], [1]):
```

Comment:
```suggestion
    if np.any((y_test != 0) & (y_test != 1)) or np.any(y_test.sum(axis=1) != 1):
```
OK I think this does it: the array is one-hot encoded if every entry is either 0 or 1 _and_ each row sums to 1. This should ensure that every row only has one 1 with all other entries being 0.